src/NLog/Layouts/XmlLayout.cs

```diff
@@ -381,38 +395,188 @@ private void AppendLogEventXmlProperties(LogEventInfo logEventInfo, StringBuilde
                     if (string.IsNullOrEmpty(key))
                         continue;
                     object propertyValue = MappedDiagnosticsLogicalContext.GetObject(key);
-                    AppendXmlPropertyValue(key, propertyValue, sb, sb.Length == orgLength);
+                    AppendXmlPropertyValue(key, propertyValue, Convert.GetTypeCode(propertyValue), sb, orgLength);
                 }
             }
 #endif
 
             if (IncludeAllProperties && logEventInfo.HasProperties)
             {
-                foreach (var property in logEventInfo.Properties)
+                var propertiesList = logEventInfo.CreateOrUpdatePropertiesInternal(true) as IEnumerable<MessageTemplates.MessageTemplateParameter>;
```

Comment:
why the `as`?

Cant we use  `IEnumerable<MessageTemplates.MessageTemplateParameter> propertiesList = logEventInfo.CreateOrUpdatePropertiesInternal(true)` (runtime safe)

Reply:
PropertiesDictionary simulates two IEnumerables. The old LogEventInfo.Properties (`IDictionary<object, object>`) and the new LogEventInfo.MessageTemplateParameters (`IEnumerable<MessageTemplateParameter>`). I want to enumerate the new one to access CaptureType.

Reply:
Yes I understand, 

but we can use `IEnumerable<MessageTemplates.MessageTemplateParameter> propertiesList = logEventInfo.CreateOrUpdatePropertiesInternal(true)` isn't?

Reply:
Yes you are right. Fixed

Reply:
🎉